src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.Fixer.cs

```diff
@@ -0,0 +1,177 @@
+
+            var fixComponents = GetFixComponents(generator, typedNodeToFix);
+
+            SyntaxNode stringArgument = generator.IdentifierName(fixComponents.StringArgumentName);
+            SyntaxNode lengthNode = generator
+                .MemberAccessExpression(
+                    stringArgument, 
+                    generator.IdentifierName(nameof(string.Length)));
+
+            var lengthArgument = generator
+                .SubtractExpression(
+                    lengthNode, 
+                    fixComponents.OriginalInnerArguments[0].Expression);
+
+            var newNode = generator.InvocationExpression(
+                fixComponents.TargetMethod, 
+                stringArgument,
+                fixComponents.OriginalInnerArguments[0], 
+                lengthArgument);
+            var newRoot = root
+                .ReplaceNode(
+                    typedNodeToFix, 
+                    newNode);
+            return document.WithSyntaxRoot(newRoot);
+        }
+
+        private struct FixComponents
+        {
+            public FixComponents(
+                string stringArgumentName, 
+                SyntaxNode targetMethod, 
+                SeparatedSyntaxList<ArgumentSyntax> originalInnerArguments)
+            {
+                StringArgumentName = stringArgumentName;
+                TargetMethod = targetMethod;
+                OriginalInnerArguments = originalInnerArguments;
+            }
+
+            public string StringArgumentName { get; }
+            public SyntaxNode TargetMethod { get; }
+            public SeparatedSyntaxList<ArgumentSyntax> OriginalInnerArguments { get; }
+        }
+
+        private static async Task<Document> FixCodeTwoParameters(
+            Document document,
+            SyntaxNode root,
+            SyntaxNode nodeToFix,
+            CancellationToken cancellationToken)
+        {
+            var typedNodeToFix = (InvocationExpressionSyntax) nodeToFix;
+            var generator = SyntaxGenerator.GetGenerator(document);
+
+            FixComponents fixComponents = GetFixComponents(generator, typedNodeToFix);
+            
+            SyntaxNode stringArgument = generator.IdentifierName(fixComponents.StringArgumentName);
+
+            var newNode = generator.InvocationExpression(
+                fixComponents.TargetMethod,
+                stringArgument,
+                fixComponents.OriginalInnerArguments[0],
+                fixComponents.OriginalInnerArguments[1]);
+            var newRoot = root.ReplaceNode(typedNodeToFix, newNode);
+            return document.WithSyntaxRoot(newRoot);
+        }
+
+        private static FixComponents GetFixComponents(
+            SyntaxGenerator generator, 
+            InvocationExpressionSyntax typedNodeToFix)
+        {
+            if (typedNodeToFix.Expression is MemberAccessExpressionSyntax memberAccessExpression
+                && memberAccessExpression.Expression is IdentifierNameSyntax identifierName)
```

Comment:
same here.  no need for language specific code.

Reply:
okay... changed line 119 to the IOperation model, so typedNodeToFix is now IOperation.
How to find the language agnostic type to cast to (so I need the language agnostic representations for InvocationExpressionSyntax - as IOperation itself is too generic), MemberAccessExpressionSyntax and IdentifierNameSyntax here, I guess.

Reply:
As mentioend, i would recommend just looking at the types you're getting back in the debugger :)   For example, you'll have IInvocationExpression.  Start looking at that, and the types you get back that, and it will help guide you toward the specific operation types you'll need to consider :)